solr/core/src/java/org/apache/solr/util/numeric/IntFloatDynamicMap.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.util.numeric;
+
+import java.util.Arrays;
+
+import com.carrotsearch.hppc.IntFloatHashMap;
+import com.carrotsearch.hppc.cursors.FloatCursor;
+import com.carrotsearch.hppc.procedures.IntFloatProcedure;
+import org.apache.lucene.util.ArrayUtil;
+
+import static org.apache.solr.util.numeric.DynamicMap.mapExpectedElements;
+import static org.apache.solr.util.numeric.DynamicMap.threshold;
+import static org.apache.solr.util.numeric.DynamicMap.useArrayBased;
+
+public class IntFloatDynamicMap {
+  private int maxSize;
+  private IntFloatHashMap hashMap;
+  private float[] keyValues;
+  private float emptyValue;
+  private int threshold;
+
+  public IntFloatDynamicMap(int expectedMaxSize, float emptyValue) {
+    this.threshold = threshold(expectedMaxSize);
+    this.maxSize = expectedMaxSize;
+    this.emptyValue = emptyValue;
+    if (useArrayBased(expectedMaxSize)) {
+      upgradeToArray();
+    } else {
+      this.hashMap = new IntFloatHashMap(mapExpectedElements(expectedMaxSize));
+    }
+  }
+
+  private void upgradeToArray() {
+    keyValues = new float[maxSize];
+    if (emptyValue != 0.0f) {
+      Arrays.fill(keyValues, emptyValue);
+    }
+    if (hashMap != null) {
+      hashMap.forEach((IntFloatProcedure) (key, value) -> keyValues[key] = value);
+      hashMap = null;
+    }
+  }
+
+  private void growBuffer(int minSize) {
+    assert keyValues != null;
+    int size = keyValues.length;
+    keyValues = ArrayUtil.grow(keyValues, minSize);
+    if (emptyValue != 0.0f) {
+      for (int i = size; i < keyValues.length; i++) {
+        keyValues[i] = emptyValue;
+      }
+    }
+  }
+
+
+  public void set(int key, float value) {
+    if (keyValues != null) {
+      if (key >= keyValues.length) {
+        growBuffer(key + 1);
+      }
+      keyValues[key] = value;
+    } else {
+      this.hashMap.put(key, value);
+      this.maxSize = Math.max(key, maxSize);
+      if (this.hashMap.size() > threshold) {
```

Comment:
why it is safer?

Reply:
Did you test and debug when the DynamicMap upgrades from a map to an array internally? I mean in debug mode step by step here. I think the map first enlarges and rehashes just before the upgrade to an array.

Let's take an example with expectedKeyMax = 500K.
this.threshold = threshold(expectedKeyMax) = 500000/64 = 7812
IntFloatHashMap initial capacity = mapExpectedElements(expectedKeyMax) = (int) (threshold(expectedKeyMax) / 0.75f) = (int) (7812 / 0.75f) = 10416
IntFloatHashMap internal threshold = ceil(initial capacity * 0.75) = ceil(10416 * 0.75) = 7812
Internally the HPPC map enlarges during a put() when its size == 7812 *before* incrementing the size.
Here the condition to upgrade to an array triggers when the map size *after* the put is > 7812, so at 7813.
So I think the map first enlarges and rehashes just before we upgrade to an array, which would be wasteful.

Also, the map internal threshold is ceil(initial capacity * 0.75), but it could be without ceil() for other implementations. To be safe wrt the float rounding, I suggested to add +1 in DynamicMap. mapExpectedElements(int).